lib/core_ext/open3.rb

```diff
@@ -0,0 +1,52 @@
+module Open3
+  # This is a combination of the functionality of Open3.capture3 and
+  # Open3.pipeline_r
+  #
+  # Open3.pipeline_r doesn't support capturing stderr, so this method takes the
+  # setup code found in `Open3#popen3`, and the block from `Open3#capture3`,
+  # and combines them into a single method.
+  #
+  # Should have the same interface as `Open3.capture3`, but supports passing in
+  # pipes.
+  def pipe_capture3(env, cmds, stdin_data: '', binmode: false, **opts)
+    # Begin excerpt from Open3.popen3
+    in_r, in_w = IO.pipe
+    opts[:in] = in_r
+    in_w.sync = true
+
+    out_r, out_w = IO.pipe
+    opts[:out] = out_w
+
+    err_r, err_w = IO.pipe
+    opts[:err] = err_w
+    # End excerpt from Open3.popen3
+
+    # Inject global ENV to all commands
+    commands   = cmds.map { |cmd| Array(cmd).unshift(env) }
+    child_ios  = [in_r, out_w, err_w]
+    parent_ios = [in_w, out_r, err_r]
+
+    pipeline_run(commands, opts, child_ios, parent_ios) do |*result|
+      # Mostly the same as the block content in Open3#capture3, except that
+      # pipeline_run returns an array of Processes that have been piped to, and
+      # we just want the status of the last one.
+      pipe_in, pipe_out, pipe_err, child_pids = result
+      # Begin block excerpt from Open3#capture3
+      if binmode
+        pipe_in.binmode
+        pipe_out.binmode
+        pipe_err.binmode
+      end
+      out_reader = Thread.new { pipe_out.read }
```

Comment:
do we need to worry about joining/closing these two threads?

Reply:
My initial reaction is "no, because I am pretty sure this code is ripped directly from Ruby core" (I basically re-arranged it), but let me double check...

Reply:
Okay, so two things:

1.  This entire section is a ripoff of [`Open3#capture3`](https://github.com/ruby/ruby/blob/0a03394b44d220e5b6573e1a9542ede6e77e3bfe/lib/open3.rb#L259-L273), only thing that is different is the last section.
2.  I recently learned this, but `Thread.value` is very similar to `Thread.join`, so that is doing the `.wait` for us.